rfd/0085-better-react-testing-remove-default-exports.md

```diff
@@ -0,0 +1,164 @@
+restrict usage of Storybook mostly to our shared component library, and test our fullscreen components another way.
+
+It's still okay to create our larger components in Storybook (if the developer wants to) but we should not be using these stories in our Jest tests.
+
+It might be worthwhile looking into something like [Chromatic](https://www.chromatic.com/) for visual UI testing.
+
+### Changing how we write tests
+
+We should be testing all the different state possibilities in Jest and React Testing Library, mocking any data or
+network requests to cause the component to render into the state we want. This ensures that the internal logic
+of our components (generally the most critical and difficult-to-get-right aspect of frontend code) gets tested,
+as well as the final UI state.
+
+Tests should check for elements that exist for the specific state being tested, i.e. an error message when the data has
+failed to load. Elements should be clicked on or events triggered in order to change the state of the component.
+
+By writing tests as described above, snapshot tests typically become redundant and therefore unnecessary. Besides the problem of skipping the business logic, snapshots consistently break for non-bugs, such as when styles are updated, and
+generally don't provide much if any security around behaviour or appearance. Further reading -
+https://medium.com/@sapegin/whats-wrong-with-snapshot-tests-37fbe20dfe8e.
+
+### Changing how we export components
+
+Default exports should be avoided except when absolutely necessary (for example, with
+`React.Lazy`), as they degrade the developer experience
+for no discernible benefit. For more fully fleshed out reasoning, refer to the articles below:
+- https://blog.neufund.org/why-we-have-banned-default-exports-and-you-should-do-the-same-d51fdc2cf2ad
+- https://rajeshnaroth.medium.com/avoid-es6-default-exports-a24142978a7a
+- https://blog.piotrnalepa.pl/2020/06/26/default-exports-vs-named-exports/
+- https://ilikekillnerds.com/2019/08/default-exports-bad/
+
+When needing to export something as default, we should export it like so:
+
+```typescript jsx
+function SomeComponent() {
+  return (
+    <div>
+      hello!
+    </div>
+  );
+}
+
+export { SomeComponent as default };
+```
+
+This provides a clearer, more explicit export.
+
+### Code Coverage
+
+Our current code coverage is:
+
+```
+=============================== Coverage summary ===============================
+Statements   : 36.12% ( 3843/10639 )
+Branches     : 39.44% ( 1484/3762 )
+Functions    : 32.36% ( 1376/4251 )
+Lines        : 36.31% ( 3759/10350 )
+================================================================================
+```
+
+We should add code coverage to CI in order to track progress. One tool we can use for this is
+[Codecov](https://about.codecov.io/), which provide a free plan for open source tools. This would provide a useful UI
```

Comment:
I would be interested in tools that can compute coverage locally rather than using a SaaS. Codecov in particular had a pretty nasty supply chain breach, so pulling in a tool like that would require a security review.

Reply:
Might be worth looking at https://github.com/marketplace/actions/code-coverage-report, and then we can use `jest --coverage` locally to get a UI view of coverage

Reply:
I've updated this to point at the suggestions above, as they do make more sense